### PR TITLE
Symlink dirs cli

### PR DESCRIPTION
**Pull request by datamel**
Adds cli option (`--symlink-dirs`) for over-riding localhost symlink dirs in global config when executing `cylc install`. 
Removes --no-symlink-dirs option, this is replaced by --symlink-dirs=None

Also moves symlink dirs into install section of config...(these changes close https://github.com/cylc/cylc-flow/issues/4219)

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.py` and
  `conda-environment.yml`.

- [x] Appropriate tests are included (unit and/or functional).
- [x] Appropriate change log entry included.
- [x] (master branch) I have opened a documentation PR at https://github.com/cylc/cylc-doc/pull/252

### CODE REVIEW

File: cylc/flow/workflow_files.py
Comment (oliver-sanders) [inline cylc/flow/workflow_files.py:1375]: Just incase users try to "unset" a symlink dir by doing `<dir>=` it might be worth swapping the empty string for `None` (which is think is handled).

```suggestion
            symdict['localhost'][key] = val.strip() or None
```

**Comment by hjoliver** [comment]
>  Removes `--no-symlink-dirs` option, this is replaced by `--symlink-dirs=None`

I'm not sure about this.  The original `--no-symlink-dirs` is pretty unambiguous and intuitive, but `--symlink-dirs=None` kind of exposes Python internals through the CLI and is less intuitive. Users would more likely guess `--symlink-dirs=""` (empty string), which does nothing.

**Comment by hjoliver** [comment]
I'm slight confused about how this works (from a quick test, without examining the code).  If I set (e.g.) `log` and `work` symlinks for `localhost` in `global.cylc`, both symlinks get created by `cylc install`.  If I then override `log` using `cylc install --symlink-dirs`, I find that `cylc install` creates only the overridden `log` symlink, and the global config `work` one gets created later by `cylc play` ??

**Review by oliver-sanders — COMMENTED** [review]
Tested, looks good:

```console
$ cylc config -i '[install][symlink dirs]'
[[[localhost]]]
    log = $HOME/cylc-log
    run = $HOME/cylc-runn
    share = 
    share/cycle = 
    work = 
$ cylc install one 
INSTALLED one/run1 from .../one
$ cylc install one --symlink-dirs=
INSTALLED one/run2 from .../one
$ cylc install one --symlink-dirs=log=$HOME/cylc-log
INSTALLED one/run3 from .../one
$ tree -l
.
├── _cylc-install
│   └── source -> .../one
│       └── suite.rc
├── run1 -> /cylc-runn/cylc-run/one/run1
│   ├── flow.cylc -> suite.rc
│   ├── log -> /cylc-log/cylc-run/one/run1/log
│   │   ├── install
│   │   │   └── 20210616T102225Z-install.log
│   │   └── version
│   │       ├── uncommitted.diff
│   │       └── vcs.conf
│   └── suite.rc
├── run2
│   ├── flow.cylc -> suite.rc
│   ├── log
│   │   ├── install
│   │   │   └── 20210616T102232Z-install.log
│   │   └── version
│   │       ├── uncommitted.diff
│   │       └── vcs.conf
│   └── suite.rc
├── run3
│   ├── flow.cylc -> suite.rc
│   ├── log -> /cylc-log/cylc-run/one/run3/log
│   │   ├── install
│   │   │   └── 20210616T102253Z-install.log
│   │   └── version
│   │       ├── uncommitted.diff
│   │       └── vcs.conf
│   └── suite.rc
└── runN -> run3  [recursive, not followed]
```


Tested again overriding run and work, the `cylc play` overriding has been addressed.

File: cylc/flow/pathutil.py
Comment (oliver-sanders) [inline cylc/flow/pathutil.py:179]: Don't need to add the types in the docstrings if using type hints.

File: cylc/flow/workflow_files.py
Comment (oliver-sanders) [inline cylc/flow/workflow_files.py:1108]: I think the `Union` isn't doing anything here:

```diff
 Optional[
-    Union[
          Dict[str, Dict[str, Any]],
         None
-    ]
 ]
```

**Comment by datamel** [comment]
> I'm slight confused about how this works (from a quick test, without examining the code). If I set (e.g.) `log` and `work` symlinks for `localhost` in `global.cylc`, both symlinks get created by `cylc install`. If I then override `log` using `cylc install --symlink-dirs`, I find that `cylc install` creates only the overridden `log` symlink, and the global config `work` one gets created later by `cylc play` ??

Sorry about that Hilary. Chatted this through with the team this end and have implemented the following changes. 
- If `cylc play`-ing an installed flow, localhost symlink dirs will now be skipped. 
- Missing symlink dirs entered on the command line will not fill from global config.
- `--symlink-dirs=""` is now the interface for skipping symlink dirs when installing flows.

The dropping of `--no-symlink-dirs` was an effort to keep the cli concise, rather than bloating with too many options.

Hope this is ok, happy to make changes if you wish.

**Review by oliver-sanders — APPROVED** [review]

File: cylc/flow/scripts/install.py
Comment (hjoliver) [inline cylc/flow/scripts/install.py:114]:
```suggestion
            "Enter a list, in the form 'log=path/to/log-dir, share=path/to/share-dir, ...'"
            ". Use this option to override local symlinks for directories run,"
            " log, work, share, share/cycle, as configured in global.cylc. "
            "Paths may contain environment variables.  Any use of this "
            "option overrides all global symlink configuration. "
            "Enter an empty list \"\" to skip making localhost symlink dirs."
```

Also, from testing, if I use a space instead of comma to separate `log=..., share=...` (e.g.),  `cylc install` silently fails to make any symlinks. We should probably support comma or space, or else explicitly fail plain spaces to avoid the parsing complication of filenames containing spaces.

Comment (hjoliver) [reply cylc/flow/scripts/install.py:114]: And regarding this (my wording):
> Any use of this option overrides all global symlink configuration.

I presume that was a deliberate choice?  What's the argument for not having individual path override?

File: cylc/flow/workflow_files.py
Comment (hjoliver) [inline cylc/flow/workflow_files.py:565]:
```suggestion
        if symlinks_created:
```
No reason to use explicit `bool()` here, is there?

File: cylc/flow/workflow_files.py
Comment (hjoliver) [inline cylc/flow/workflow_files.py:593]:
```suggestion
    return cylc_install_dir.is_dir() or alt_cylc_install_dir.is_dir():
```

File: cylc/flow/workflow_files.py
Comment (hjoliver) [inline cylc/flow/workflow_files.py:1165]:
```suggestion
    if symlinks_created:
```

File: cylc/flow/workflow_files.py
Comment (hjoliver) [inline cylc/flow/workflow_files.py:1377]: (See comment above on accidental (or assumed) use of spaces instead of commas).

File: cylc/flow/workflow_files.py
Comment (hjoliver) [inline cylc/flow/workflow_files.py:1382]: Better to fail this as a command line error (it's just a command, so can be redone easily with no typos in it)?

**Review by hjoliver — COMMENTED** [review]
Looks great, tests as working, just a few minor suggestions and questions.

File: cylc/flow/scripts/install.py
Comment (datamel) [reply cylc/flow/scripts/install.py:114]: This was a request from @dpmatthews. We discussed did discuss other options here, with the  suggestion from @oliver-sanders that we could add a further option in time, to fill links from config if users wanted it.

Comment (dpmatthews) [reply cylc/flow/scripts/install.py:114]: My reasoning is that this way it's clear to the user what they will get. Otherwise, if I wanted to configure my workflow to use a single filesystem I'd need to either know the global config setup to know what to override or I'd need to set all the possible symlink settings just to be sure. We could support some method of choosing whether to do the individual path override but that seems like unnecessary complexity for an option which we don't expect to be widely used.

File: cylc/flow/workflow_files.py
Comment (datamel) [reply cylc/flow/workflow_files.py]: I have implemented explicitly failing plain spaces, due to the complications with file names containing spaces.

**Review by MetRonnie — COMMENTED** [review]
I think this is the PR you wanted me to check after #4237 was merged

File: cylc/flow/pathutil.py
Comment (datamel) [reply cylc/flow/pathutil.py:219]: Line 226 has the the return false.

File: tests/unit/conftest.py
Comment (datamel) [reply tests/unit/conftest.py:68]: oopsie, thanks

File: cylc/flow/pathutil.py
Comment (MetRonnie) [reply cylc/flow/pathutil.py:219]: But this checks if the path exists _after_ the target is created.
1. When the `~/cylc-run/foo/share -> ~/sym-share/cylc-run/foo/share` symlink & target already exist, and we're now on `share/cycle`
2. The line `target.mkdir(...)` above will create `~/sym-share/cylc-run/foo/share/cycle`.
3. At this point `path` exists now - `path` is `~/cylc-run/foo/share/cycle`, but because `~/cylc-run/foo/share` is a symlink, `path` is actually  `~/sym-share/cylc-run/foo/share/cycle`

(Sorry, the comment in the code doesn't give a good explanation I now realise! And I should have created a test case for `share` and `share/cycle` being the same)

File: cylc/flow/workflow_files.py
Comment (MetRonnie) [inline cylc/flow/workflow_files.py:1505]: Suggest calling this something like `parse_comma_sep_symlink_dirs` to avoid confusion with `get_symlink_dirs`

File: cylc/flow/workflow_files.py
Comment (MetRonnie) [inline cylc/flow/workflow_files.py:1539]: Rather than rewriting these strings,
```suggestion
        possible_symlink_dirs = WorkflowFiles.SYMLINK_DIRS.symmetric_difference(
            {'', WorkflowFiles.RUN_DIR}
        ).
        if key not in possible_symlink_dirs:
```
But it is a bit awkward due to the use of `''` instead of `'run'` in `WorkflowFiles.SYMLINK_DIRS`.

(If you do use this suggestion, instantiate `possible_symlink_dirs` outside this `for` loop rather than where I've got it)

https://docs.python.org/3.7/library/stdtypes.html#frozenset.symmetric_difference

---

Alternative:
```suggestion
        if key not in {WorkflowFiles.RUN_DIR, WorkflowFiles.LOG_DIR, ...} # etc
```

File: tests/unit/conftest.py
Comment (MetRonnie) [inline tests/unit/conftest.py:68]:
```suggestion
        installed: If True, make it look like the workflow was installed
            using cylc install (creates _cylc-install dir).
        named: If True and installed is True, the _cylc-install dir will
            be created in the parent to make it look like this is a
            named run.
```

File: tests/unit/conftest.py
Comment (MetRonnie) [inline tests/unit/conftest.py:83]:
```suggestion
                if named:
                    if len(Path(reg).parts) < 2:
                        raise ValueError("Named run requires two-level reg")
```

File: tests/unit/conftest.py
Comment (MetRonnie) [inline tests/unit/conftest.py:73]:
```suggestion
        installed: bool = False,
        named: bool = False
    ) -> Path:
```

File: cylc/flow/pathutil.py
Comment (datamel) [reply cylc/flow/pathutil.py:218]: No, it is not redundant... consider the case where a user sets up a symlink, different to that of what is in the `global.cylc [install] [symlink dirs]` section and then runs `cylc play`. In this scenario...
- `path.exists()` returns True
- `path.is_symlink()` returns True
- `path.samefile(target)` results in error with traceback, since that is the default for `samefile` if either path or target do not exist.
In this case, we want the skipping to be logged and it to pass through without error.

**Review by hjoliver — APPROVED** [review]
Approving this now, assuming @MetRonnie's final comments will be addressed (or countered). Very nice, thanks @datamel :+1:

File: cylc/flow/pathutil.py
Comment (MetRonnie) [reply cylc/flow/pathutil.py:218]: Ah ok, might be worth a comment in the code

Comment (datamel) [reply cylc/flow/pathutil.py:220]: Thanks, comment added.

File: tests/unit/test_workflow_files.py
Comment (MetRonnie) [inline tests/unit/test_workflow_files.py:1485]: 🧐
```suggestion
    if err_msg is not None:
```

**Review by MetRonnie — APPROVED** [review]
Happy with the response to my review

**Pull request merged**
